// node-red-contrib-temp/temp.js
module.exports = function(RED) {
    // function Compareeq(a, b) {
    //     return (a == b) ? true : false;
    // }
    function Comparegt(a, b) {
        return (a > b) ? true : false;
    }
    // function Comparegte(a, b) {
    //     return (a >= b) ? true : false;
    // }
    // function Comparelt(a, b) {
    //     return (a < b) ? true : false;
    // }
    function Comparelte(a, b) {
        return (a <= b) ? true : false;
    }
    function CheckNode(config) {
        RED.nodes.createNode(this,config);
        this.temp = config.temp;
        this.compare = config.compare;

        var node = this;

        node.on('input', function(msg) {
            var inputJson = JSON.parse(msg.payload)
            var strType = "NULL"
            if(inputJson.length>0){
                strType = inputJson[0].bn
            }
            if (typeof strType !== 'undefined' && strType!="NULL"){
                var typeLog = strType.split("_")[2] + strType.split("_")[3]
                if (typeLog==="OPUSTATUS"){
                    var tmp =  inputJson[1].v
                    var mcu_id = inputJson[0].vs
                    switch (node.compare) {
                        // case "eq" :
                        //     if(Compareeq(vl, node.volume)) {
                        //         msg.payload = "";
                        //         node.send(msg);
                        //     } else {
                        //         node.warn("Not equal");
                        //     }
                        //     break;
                        case "gt":
                            if(Comparegt(tmp, node.temp)) {
                                msg.payload = "Loa "+mcu_id+" có nhiệt độ vượt ngưỡng cho phép!";
                                node.send(msg);
                            } 
                            // else {
                            //     node.warn("");
                            // }
                            break;
                        // case "gte":
                        //     if(Comparegte(vl, node.volume)) {
                        //         msg.payload = true;
                        //         node.send(msg);
                        //     } else {
                        //         node.warn("Not greater than or equal");
                        //     }
                        //     break;
                        // case "lt":
                        //     if(Comparelt(vl, node.volume)) {
                        //         msg.payload = true;
                        //         node.send(msg);
                        //     } else {
                        //         node.warn("Not less than");
                        //     }
                        //     break;
                        case "lte":
                            if(Comparelte(tmp, node.temp)) {
                                msg.payload = "Loa "+mcu_id+" có nhiệt độ nhỏ hơn ngưỡng cho phép!";
                                node.send(msg);
                            } 
                            // else {
                            //     node.warn("Not less than or equal");
                            // }
                            break;
                    }
                }
            }

   

        });
    }
    RED.nodes.registerType("anomaly temp TTTM",CheckNode);
}
